### ISSUE TITLE
High impact bug: retainOnDelete on dynamoDB table does not preserve the table!

### ISSUE THREAD

**Issue by Jimmy89** [pulumi/pulumi#13428]
### What happened?

Luckily this only happened to a test data table, but like the title states: deleting a table with `retainOnDelete` set to `true` on a DynmoDB table does _NOT_ preverse the table as it should. Deleting the resource in pulumi deletes the table as well.

### Expected Behavior

Not deleting a table full of data when retainOnDelete is set.

### Steps to reproduce

```
const dynamodbTable = new aws.dynamodb.Table(`dynamodb-test-table`, {
  attributeDefinitions: [
    {
      attributeName: "PK",
      attributeType: "S",
    },
    {
      attributeName: "SK",
      attributeType: "S",
    },
    {
      attributeName: "G1PK",
      attributeType: "S",
    },
    {
      attributeName: "G1SK",
      attributeType: "S",
    },
    {
      attributeName: "G2PK",
      attributeType: "S",
    },
    {
      attributeName: "G2SK",
      attributeType: "S",
    },
  ],
  keySchema: [
    {
      attributeName: "PK",
      keyType: "HASH",
    },
    {
      attributeName: "SK",
      keyType: "RANGE",
    },
  ],
  globalSecondaryIndexes: [
    {
      indexName: "GSI1",
      keySchema: [
        {
          attributeName: "G1PK",
          keyType: "HASH",
        },
        {
          attributeName: "G1SK",
          keyType: "RANGE",
        },
      ],
      projection: {
        projectionType: "INCLUDE",
        nonKeyAttributes: ["Id"],
      },
    },
    {
      indexName: "GSI2",
      keySchema: [
        {
          attributeName: "G2PK",
          keyType: "HASH",
        },
        {
          attributeName: "G2SK",
          keyType: "RANGE",
        },
      ],
      projection: {
        projectionType: "INCLUDE",
        nonKeyAttributes: ["St"],
      },
    },
  ],
  pointInTimeRecoverySpecification: {
    pointInTimeRecoveryEnabled: true,
  },
  sSESpecification: {
    sSEEnabled: true,
  },
  streamSpecification: {
    streamViewType: "KEYS_ONLY"
  },
  tableName: "test-table",
  billingMode: "PAY_PER_REQUEST",
  timeToLiveSpecification: {
    attributeName: "TTL",
    enabled: false,
  },
  tags,
}, { protect: false, replaceOnChanges: ["tableName"], deleteBeforeReplace: true, retainOnDelete: true })

```

Create and delete the resource by renaming the table to `test-table-second`.

### Output of `pulumi about`

```
NAME                VERSION
@pulumi/aws         5.41.0
@pulumi/aws-native  0.67.0
@pulumi/pulumi      3.74.0
@pulumi/random      4.13.2
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Jimmy89** [comment]
@Frassle sorry, I did not see your comment.  Yes I did, I set `retainOnDelete` to true and run the update in the same run. I was expecting that it would take precedence as it is a pulumi property.

**Comment by Frassle** [comment]
Yup that was my suspicion this is technically by design, but when this was raised we did have another talk internally if that was the _right_ design.

Currently the system behaves as follows, Option A:
| Operation | RetainOnDelete in state | RetainOnDelete from program | Is retained? | 
| --- | --- | --- | --- |
| Delete | Yes | N/A (A plain delete happens because the resource is no longer sent by the program) | Yes |
| Delete | No| N/A | No |
| Replace | Yes | Yes | Yes |
| Replace | Yes | No | Yes |
| Replace | No | Yes | No |
| Replace | No | No | No |

That is the decision of if the resource should be deleted or not is entirely governed by what's currently in state. Setting the option just configures the state for the _next time_ an update is run.
This means in your case where you want to replace a resource but also retain it you need to do two updates, first to set RetainOnDelete, then to do the replace.

We could change this to the following, Option B:
| Operation | RetainOnDelete in state | RetainOnDelete from program | Is retained? | 
| --- | --- | --- | --- |
| Delete | Yes | N/A | Yes |
| Delete | No| N/A | No |
| Replace | Yes | Yes | Yes |
| Replace | Yes | No | Yes |
| Replace | No | Yes | Yes |
| Replace | No | No | No |

This is the same in all cases _except_ for when doing a replace we'll also look at what the new retain on delete setting is and if it's true we'll retain the old resource that's being replaced.
This means in cases where you want to replace a resource and then retain future copies of it you would now need to do two updates, first to do the replace deleting the old resource, then to just set the RetainOnDelete flag.

Alternatively we could fully prefer program input, option C:
| Operation | RetainOnDelete in state | RetainOnDelete from program | Is retained? | 
| --- | --- | --- | --- |
| Delete | Yes | N/A | Yes |
| Delete | No| N/A | No |
| Replace | Yes | Yes | Yes |
| Replace | Yes | No | No |
| Replace | No | Yes | Yes |
| Replace | No | No | No  |

This ignores the flag saved to state when doing a replace instead using the program input. Again it means if you want to replace a resource and then retain future copies you'd need to do two updates (as above), but it also means that if you want to retain the current copy but not retain future ones after a replace you'd also need to do two updates. One to trigger the replace while RetainOnDelete was still true, then another to set it to false.

So either way has at least one case where two updates are required, it's just a question of which one is "right". We'd gone with the current behavior because we thought it was probably more intuitive to think of RetainOnDelete as a setting on state of the old resource, while the program description was for the new resource. 

But I think option B has the advantage that makes it very hard to accidentally trigger a delete. A and C both have cases where you might think the option was going to be used (either because it true in state or from the program) but then the resource is deleted anyway.

We'll discuss this internally again, but very open to community comments on this as well.

**Comment by Jimmy89** [comment]
Thank you for the extensive feedback.

The way I see it as a user is: when I run an updated version of a program I expect pulumi to update the infrastructure in a single run _with the settings that are true within the context_. Like if I add an attribute to the `ignoreChanges` key, I expect the key to be ruled out at the first time I run the program. I understand that the `protect` key is an extra protection measurement that needs to be 'saved' again before being able to for example delete a resource. The retainOnDelete I expect to run on the first run, just like ignoreChanges, as well, it's quite crucial (in fact I would argue just as important as 'protect: true`). So definitely something pulumi should prioritize before apply updates to any attribute of the attached resource.

I would argue:
Program: protect: true --> highest priority for a pulumi run, first thing you save before applying any other update
Program: protect: false --> needs a second run
Program: retainOnDelete: true --> highest priority for pulumi run, first thing you save/go through before applying an update or deleting a resource.
Program: retainOnDelete: false (state: true) --> needs a second run.

Even better would be if pulumi would run/plan twice (secretly):
First do a round of updates on the pulumi resource changes.
Then execute again the actions that must be triggered on the pulumi resources.

**Comment by humbleice** [comment]
I've also found these options confusing and somewhat difficult to work with.  My concern is mostly around code integration and deployments.  As a developer is not intuitive nor efficient to have to make a state change on a resource, then get those changes approved, merged, and deployed in all environments before making another round of changes on the same resource after the state has changed.  If those separate code changes accidentally get merged and go out as a single deployment (versus separately), the final results will be different.

> Even better would be if pulumi would run/plan twice (secretly)

Something like this might be nice.  Perhaps there could be ways to configure how pulumi runs, particularly related to things like order of updating state.

**Comment by Frassle** [comment]
> > Even better would be if pulumi would run/plan twice (secretly)
>
> Something like this might be nice. 

I don't think this would help anything. e.g. if your program has changed such that resource R needs a replace then you might want the current state retained (i.e. not deleted, but still do the replace) or you might want the new state retained. There is no way to perfectly express both those options in a program with a single goal state. Running the program twice doesn't get us any new information to make decisions from.

Whatever variant of the above we chose there would be some scenario where the only way to express what you want would be to do two updates.

I think the current behaviour is reasonable. The data saved to state defines what we do with the old version, the data in the goal decides what we do with the new version.
So if your state has retain=false and then you do a replace with retain=true that means just delete the old resource but then mark the new resource as not to be deleted.

Again if we always used the data from the program, so doing a replace and setting retain at the same time would retain the old resource, then if you wanted to express the change of "actually delete the old state then mark my new resource as retained" you'd have to do that in two updates.

So some scenarios are always going to need two updates. I think one thing we could add here to make it a little easier is to add `pulumi state protect/retain/unretain` like we have `pulumi state unprotect`. Then if you wanted to replace a resource but retain the old one you could run `pulumi state retain X` then run the update.

**Comment by maroshvarchola** [comment]
Hello! Bringing this up again. I would totally aggree that current behaviour is ok, thought, sometimes it would be very efficient to add ability of Option B or C, something like "forcing" for the current run. For our automation it would be really nice, currently we are solving renaming of github repository. Currently, when rpeo name is changed, it wants to replace branches, branch protections, default branch etc... which is not quite correct.

Or put this in another way/side: How do you manage renaming repository in easy way using pulumi, when you manage also branches, protections, default branches etc?

EDIT: Or maybe even better would be to respect forcefully `"ignoreOnChanges": ["repository"]` (but this is example in my use case)
